### PR TITLE
refactor pod lookup code

### PR DESCRIPTION
**Pull request by ramaraochavali**
The pod lookup exists in endpoint and endpoint slice - This PR refactors it. Also there was a bug in key used while getting pod from shared informer (was using endpoint key instead of pod key) - fixed that. Renamed some functions.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go [hunk 1]
```diff
@@ -116,3 +118,31 @@ func updateEDS(c *Controller, epc kubeEndpointsController, ep interface{}, event
...
 }
+
+func getPod(c *Controller, ip string, ep *metav1.ObjectMeta, targetRef *v1.ObjectReference, host host.Name) *v1.Pod {
+	pod := c.pods.getPodByIP(ip)
+	if pod != nil {
+		return pod
+	}
+	// This means, the endpoint event has arrived before pod event.
+	// This might happen because PodCache is eventually consistent.
+	if targetRef != nil && targetRef.Kind == "Pod" {
```
Comment (ramaraochavali) [inline pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go:129]: @howardjohn please review keys closely

File: pilot/pkg/serviceregistry/kube/controller/pod.go [hunk 2]
```diff
@@ -44,21 +40,23 @@ type PodCache struct {
 	// pod cache if a pod changes IP.
 	IPByPods map[string]string
 
-	// map of IP to endpoint names
-	needResync     map[string]sets.Set
-	endpointUpdate func(string)
+	// needResync is map of IP to endpoint names. This is used to requeue endpoint
+	// events when pod event comes. This typically happens when pod is not available
+	// in podCache when endpoint event comes.
+	needResync         map[string]sets.Set
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/pod.go:46]: not sure why the value is also a map?

File: pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go [hunk 3]
```diff
@@ -116,3 +118,31 @@ func updateEDS(c *Controller, epc kubeEndpointsController, ep interface{}, event
...
+		key := kube.KeyFunc(targetRef.Name, targetRef.Namespace)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go:130]: This looks right but I don't understand how the test passes with and without this change?

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go:130]: Yes. I will check the problem with the test

**Pull request merged**